File: c7_binaryTree/L437.cpp
//
// Created by 李恒 on 2020/7/14.
//
#include <iostream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution{
public:
    //在递归中又嵌套了一个递归
    int pathSum(TreeNode* root , int sum){
        if(root == NULL)
            return 0 ;
        // 对于每个节点考虑三种情况：1.路径包含这个节点 2.路径不包含这个节点，以当前节点的左子节点开头
        // 3.路径不包含这个节点，以当前节点的右子节点开头
        int res = findPath(root , sum );
        res += pathSum(root->left , sum );
        res += pathSum(root->right , sum );
        return res ;
    }

private:
    // 在以node为根节点的二叉树中寻找包含node的路径，和为sum 返回路径个数
    // 这里要注意的是，路径并不是以叶子节点结尾，所以写法逻辑和前面的案例不一样
    int findPath(TreeNode* node , int sum){
        if(node == NULL)
            return 0 ;
        int res = 0 ;
        if(node->val == sum )
            res += 1 ;
        res +=  findPath(node->left , sum - node->val);
        res +=  findPath(node->right , sum - node->val);
        return res ;
    }
};

File: python-algorithm/ds_5_tree/tree_1_4_path_recurrent/L437.py
#--coding:utf-8--


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 和判断子树572那道题的结构很类似

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if root == None:
            return 0
        # 对于每个节点考虑三种情况：
        # 1.路径包含这个节点
        # 2.路径不包含这个节点，以当前节点的左子节点开头
        # 3.路径不包含这个节点，以当前节点的右子节点开头
        res = self.start_of_curnode(root, sum)
        res += self.pathSum(root.left, sum)
        res += self.pathSum(root.right, sum)
        return res

    # 返回以当前节点为根节点的子树的路径和我sum的个数
    def start_of_curnode(self, root, sum):
        if root == None:
            return 0
        res = 0
        if root.val == sum:
            res += 1
        res += self.start_of_curnode(root.left, sum-root.val)
        res += self.start_of_curnode(root.right, sum-root.val)
        return res
